types/react-redux/index.d.ts

```diff
@@ -73,9 +73,9 @@ type Shared<
 // Will not pass through the injected props if they are passed in during
 // render. Also adds new prop requirements from TNeedsProps.
 export interface InferableComponentEnhancerWithProps<TInjectedProps, TNeedsProps> {
-	(
-		component: StatelessComponent<TInjectedProps>
-	): ComponentClass<TNeedsProps> & {WrappedComponent: StatelessComponent<TInjectedProps>}
+    <P extends Shared<TInjectedProps, P>>(
+		component: StatelessComponent<P>
+	): ComponentClass<Omit<P, keyof Shared<TInjectedProps, P>> & TNeedsProps> & {WrappedComponent: StatelessComponent<P>}
 	<P extends Shared<TInjectedProps, P>>(
```

Comment:
Why is there a separate definition for `StatelessComponent` vs `ComponentType`? `ComponentType` is a superset of `StatelessComponent`.

It looks like we should just be able to remove this overloaded definition.

Reply:
I found the root of the issue. Basically, in commit https://github.com/DefinitelyTyped/DefinitelyTyped/pull/25872 there was introduced bidirectional type inference which couldn't work. i.e. for the `StatelessComponent`, the ` based on the connect function, while in the `InferableComponentEnhancerWithProps` was inferred based on connect, while in other cases it was inferred based on the connected component.